### PR TITLE
feat(amd): add gluon DSV4.1 CSA2 indexer

### PR DESCRIPTION
**Pull request by Yu-Zhewen**
## Summary

- Add CSA2 `dsv41.index_topk` Gluon on gfx950 and gfx1250: score page-planar MXFP4 index-K with scaled MFMA / WMMA, then select Top512 rows and optional Top2048 blocks with ATen.
- Keep Gluon Full through the model's 1M context by bounding FP32 logits to 32 MiB per score query tile (up to 256 queries at 32K rows, 64 at 128K, and 8 at 1M). Reindex scores only its at-most-16K candidate rows.
- Preserve gapped arena page strides without copying ordinary page bytes; normalize only non-unit inner byte strides, and enforce visible/table/page bounds before selection.
- Register 1..32 local or replicated MXFP4 index heads as `SPECIALIZED` with `native_indexer=False`. TP-sharded heads, 132-byte FP8 rows, and larger head counts retain portable Triton; Blackwell/Hopper native selection is unchanged.

## Main vs this PR

Serving calls `dsv41.index_topk(..., solution=None)` explicitly. On AMD `main` has no Gluon indexer, so `select_kernel` picks PORTABLE `triton_dsv41_index_topk`, which scans 16 partitions and merges their partial TopK. This PR selects `gluon_dsv41_index_topk_gfx950` or `gluon_dsv41_index_topk_gfx1250` for the Flash geometry.

## Kernel performance

Median GPU-event time vs portable `solution="triton"` after warmup. Flash TP4 serving shapes: 32 replicated indexer heads, topk 512, candidate block size 8, `query_chunk_size=256`. Decode `T=32` matches `--max-num-seqs 32`; prefill uses the host's maximum indexer query tile `T=256`. Full uses `candidate_topk=2048`; Reindex scores 2048 blocks (16384 rows).

### Decode (`dsv41.index_topk`)

| GPU | Shape | Main | PR | Speedup |
|---|---|---:|---:|---:|
| MI355X | T=32, vis=2K Full | 0.523 ms | 0.407 ms | 1.28x |
| MI355X | T=32, vis=8K Full | 0.659 ms | 0.398 ms | 1.65x |
| MI355X | T=32, vis=32K Full | 1.389 ms | 0.446 ms | 3.12x |
| MI355X | T=32, vis=64K Full | 2.393 ms | 0.680 ms | 3.52x |
| MI355X | T=32, vis=128K Full | 4.368 ms | 1.017 ms | 4.30x |
| MI355X | T=32, vis=1M Full | 32.197 ms | 7.386 ms | 4.36x |
| MI355X | T=32, vis=32K Reindex | 0.353 ms | 0.318 ms | 1.11x |
| MI455X | T=32, vis=2K Full | 0.605 ms | 0.409 ms | 1.48x |
| MI455X | T=32, vis=8K Full | 0.778 ms | 0.415 ms | 1.88x |
| MI455X | T=32, vis=32K Full | 1.852 ms | 0.453 ms | 4.09x |
| MI455X | T=32, vis=64K Full | 3.477 ms | 0.955 ms | 3.64x |
| MI455X | T=32, vis=128K Full | 6.564 ms | 1.682 ms | 3.90x |
| MI455X | T=32, vis=1M Full | 49.869 ms | 12.662 ms | 3.94x |
| MI455X | T=32, vis=32K Reindex | 0.384 ms | 0.311 ms | 1.23x |

### Prefill (`dsv41.index_topk`)

| GPU | Shape | Main | PR | Speedup |
|---|---|---:|---:|---:|
| MI355X | T=256, vis=8K Full | 2.940 ms | 0.621 ms | 4.74x |
| MI355X | T=256, vis=64K Full | 16.726 ms | 3.322 ms | 5.04x |
| MI355X | T=256, vis=128K Full | 32.449 ms | 6.564 ms | 4.94x |
| MI355X | T=256, vis=1M Full | 253.694 ms | 58.053 ms | 4.37x |
| MI355X | T=256, vis=32K Reindex | 1.777 ms | 0.881 ms | 2.02x |
| MI455X | T=256, vis=8K Full | 4.571 ms | 0.599 ms | 7.63x |
| MI455X | T=256, vis=64K Full | 25.487 ms | 6.070 ms | 4.20x |
| MI455X | T=256, vis=128K Full | 49.685 ms | 12.101 ms | 4.11x |
| MI455X | T=256, vis=1M Full | 392.367 ms | 100.914 ms | 3.89x |
| MI455X | T=256, vis=32K Reindex | 2.234 ms | 1.020 ms | 2.19x |

Gluon scores are oracle-close, not bitwise vs Triton (`rtol`/`atol` 0.1 on the graph oracle). Functional coverage includes tiled graph replay, Full above 32K, invalid/strided arena pages, MI355X, physical MI455X, and the MI450 functional simulator.

## Model performance (MI355X)

DeepSeek-V4.1-Flash TP4, AMD CI serve flags: `--max-model-len 1048576 --max-num-seqs 32 --chunked-prefill-size 8192 --max-cudagraph-capture-size 32 --disable-prefill-graph` (no `--enforce-eager`). Wall time after warmup. The baseline forces `TOKENSPEED_KERNEL_OVERRIDE_ATTENTION_DSV41_INDEX_TOPK=triton_dsv41_index_topk`; PR uses automatic Gluon selection.

`T=32` is `--max-num-seqs 32`. Prefill is 32 unique cold prompts. Decode is 32 concurrent prefix-hit generates (`n=48`) at that context.

| | Main (triton indexer) | PR (gluon indexer) | Speedup |
|---|---:|---:|---:|
| Prefill T=32, vis≈8K (8091 tok) | 32.5 s | 26.9 s | 1.21x |
| Decode T=32, vis≈8K | 52.2 ms/tok | 51.1 ms/tok | 1.02x |
| Decode T=32, vis≈32K (32362 tok) | 119.3 ms/tok | 82.3 ms/tok | 1.45x |

8K decode is still MoE-bound, so the indexer barely moves end-to-end latency. At 32K it becomes material. Unique T=32 vis≈32K main prefill is omitted because Triton first-compilation and SMG health timeouts prevented a usable baseline wall time. These serving numbers predate the subsequent arena-stride preservation fix and are therefore conservative for the PR path; the long-Full kernel rows above are from the current branch.

### CODE REVIEW

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/attention/dsv41/indexer.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    score_chunk_size,
+    process_group,
+    out,
+):
+    """GFX950 CSA2 indexer with on-device partitioned TopK merge."""
+    from tokenspeed_kernel.ops.attention.dsv41.triton import (
+        _index_gather_heads,
+        _index_scan_kernel,
+        _index_topk_outputs,
+        index_q_quantize,
```
[the remark this answers is not in the document]
Comment (Yu-Zhewen) [reply tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx950/attention/dsv41/indexer.py]: Moved CSA2 scan, gather, and quantize orchestration into `tokenspeed_kernel/.../dsv41/gluon.py` in 721da4f5. The AMD package now only exports the fused finish kernel (`dsv41_index_finish`).

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/dsv41/__init__.py [hunk 2]
```diff
@@ -367,6 +367,7 @@ def index_topk(
     score_chunk_size: int,
     process_group: torch.distributed.ProcessGroup | None,
     out: tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor] | None,
+    solution: str | None = None,
```
[the remark this answers is not in the document]
Comment (Yu-Zhewen) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/dsv41/__init__.py]: Leaving `solution: str | None = None` to match the rest of this family (`dsv4_prefill`, MLA, `selected_attention`). Runtime callers rely on automatic dispatch; forcing every site to pass `solution=None` would churn serving without changing behavior. Tests that need a specific kernel already pass `solution=` explicitly.

File: tokenspeed-kernel/test/ops/test_attention_dsv41_index_scan.py [hunk 3]
```diff
@@ -207,9 +233,11 @@ def tp_group():
         torch.distributed.destroy_process_group()
 
 
+@pytest.mark.parametrize("solution", ["triton", "gluon"])
```
[the remark this answers is not in the document]
Comment (Yu-Zhewen) [reply tokenspeed-kernel/test/ops/test_attention_dsv41_index_scan.py]: Portable CSA2 tests now pin Triton. Gluon graph-oracle and full/reindex coverage lives in `test/amd/ops/attention/test_gluon_dsv41_index_topk.py` (721da4f5).

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/dsv41/gluon.py [hunk 4]
```diff
@@ -40,6 +45,10 @@
     )
 
     _SIGNATURES = frozenset({format_signature(x=dense_tensor_format(torch.bfloat16))})
+    _INDEX_TRAITS = {
+        "native_indexer": frozenset({False}),
+        "index_heads": frozenset(range(1, 33)),
+    }
```
[the remark this answers is not in the document]
Comment (Yu-Zhewen) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/dsv41/gluon.py:53]: Fixed in 73e5bf3d. Public dispatch now carries `index_k_format`; AMD Gluon declares only `mxfp4`, so 132-byte `index_v4` rows select portable Triton. The Gluon host also rejects non-68-byte rows defensively, and dispatch/scoring regressions cover the FP8 format.

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/dsv41/_gluon/indexer.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+        # The scorers address adjacent page bytes, so normalize that layout only.
+        cache_2d = cache_2d.contiguous()
+    query_chunk_size = min(int(query_chunk_size), 256)
+    make_blocks = bool(candidate_topk)
+
+    for start in range(0, tokens, query_chunk_size):
+        end = min(start + query_chunk_size, tokens)
+        q, w, _shards = _index_gather_heads(
+            index_q[start:end], weights[start:end], process_group
+        )
```
[the remark this answers is not in the document]
Comment (Yu-Zhewen) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/dsv41/_gluon/indexer.py:175]: Fixed in 73e5bf3d by narrowing Gluon to `index_shards=1`. TP-sharded calls now select portable Triton, which preserves the documented per-shard BF16 rounding; the production replicated-head path remains Gluon. Added a 4-shard/32-gathered-head dispatch regression.

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/dsv41/_gluon/indexer.py [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+    weights,
+    index_cache,
+    page_table,
+    visible_lens,
+    candidate_blocks,
+    topk,
+    candidate_topk,
+    candidate_block_size,
+    query_chunk_size,
+    score_chunk_size,
```
[the remark this answers is not in the document]
Comment (Yu-Zhewen) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/dsv41/_gluon/indexer.py:120]: Fixed in e1153cde. `score_chunk_size` is now plumbed through the shared adapter into both scorer grids and capped at the tuned 256-row maximum. Sub-32 bounds use a rounded hardware tile with live-row store masks, avoiding overlap between adjacent chunks. Tests cover requested chunks 8/64/512 and the 8-row partial-block case on MI355X, MI455X, and the MI450 simulator.

File: tokenspeed-kernel/test/ops/test_attention_dsv41_index_scan.py [hunk 7]
```diff
@@ -72,7 +72,7 @@ def _oracle(q, weights, keys, shards):
     return scores.to(weights.dtype).float()
 
 
-def _check(ids, lengths, scores):
+def _check(ids, lengths, scores, rtol=0, atol=0):
```
[the remark this answers is not in the document]
Comment (Yu-Zhewen) [reply tokenspeed-kernel/test/ops/test_attention_dsv41_index_scan.py]: Fixed in e1153cde. `_check` now requires explicit `rtol` and `atol`; all existing call sites already provide both.

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx1250/attention/dsv41/indexer.py [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+        )
+        gl.store(
+            logits + token * logits_stride + store_pos,
+            gl.where(live, scores, -float("inf")),
+            mask=store_pos < max_candidates,
+        )
+
+
+def dsv41_index_logits_gfx1250(q, w, cache_2d, table, visible, candidates, logits):
+    """Score prepared 32-head BF16 queries into caller-owned CSA2 logits."""
```
[the remark this answers is not in the document]
Comment (Yu-Zhewen) [reply tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/gfx1250/attention/dsv41/indexer.py]: Fixed in e1153cde. Both exported gfx950/gfx1250 scorers now document every prepared input, the caller-owned `logits` initialization/mutation contract, the score-chunk bound, and the `None` return behavior.

File: tokenspeed-kernel/test/ops/test_attention_dsv41_index_scan.py [hunk 9]
```diff
@@ -207,9 +233,10 @@ def tp_group():
         torch.distributed.destroy_process_group()
 
 
-@pytest.mark.parametrize("shards", [1, 4], ids=["local", "tp4"])
-@pytest.mark.parametrize("dtype", [torch.bfloat16, torch.float32])
-def test_index_scan_graph_oracle(device, shards, dtype, tp_group):
+def run_index_scan_graph_oracle(
+    device, shards, dtype, tp_group, require, solution, rtol=0, atol=0
```
[the remark this answers is not in the document]
Comment (Yu-Zhewen) [reply tokenspeed-kernel/test/ops/test_attention_dsv41_index_scan.py]: Fixed in 17d3a504. `run_index_scan_graph_oracle` now requires explicit `rtol` and `atol`; the portable caller passes `rtol=0, atol=0`, while the Gluon caller retains its explicit 0.1 tolerances.

**Review by borontion — APPROVED** [review]
